### PR TITLE
EC/ROCM: add host execution capability

### PR DESCRIPTION
**Pull request by edgargabriel**
## What
Introduce the ability to use host based reduction and copy operations

## Why?
This avoids the cost of a kernel launch., which can be beneficial for short messages and when trying to overlap communication and computation occurring on the GPU.

### CODE REVIEW

File: src/components/ec/rocm/ec_rocm_host_operations.c [hunk 1]
```diff
@@ -0,0 +1,351 @@
+/**
```
[the remark this answers is not in the document]
Comment (edgargabriel) [reply src/components/ec/rocm/ec_rocm_host_operations.c:1]: I will give it a try, thanks for the hint!

**Comment by edgargabriel** [comment]
@vspetrov I implemented your suggest approach, and it works fundamentally. I am running however into some issues in the finalize stage. I *think* the ec/cpu is removed before ec/rocm, and because of the ee_executor_finalize() (or ee_executor_stop() ) in ec_rocm_finalize() fails, to be more precise this condition:
```
UCC_CHECK_EC_AVAILABLE(executor->ee_type);
```
Note, the condition passed during initialization when creating the cpu_executor. 
Consequently, I get some warnings at the end along the lines of 
```
 mpool.c:54   UCX  WARN  object 0x555d77922f80 was not returned to mpool ec cpu executors
```
which I think is a direct consequence of this, since the mpool_put() of the executor which is done during ee_executor_finalize() is not performed.

Do you have a suggestion on how to deal with this?

**Comment by vspetrov** [comment]
> @vspetrov I implemented your suggest approach, and it works fundamentally. I am running however into some issues in the finalize stage. I _think_ the ec/cpu is removed before ec/rocm, and because of the ee_executor_finalize() (or ee_executor_stop() ) on the cpu_executor that I call in ec_rocm_finalize() fails, to be more precise this condition:
> 
> ```
> UCC_CHECK_EC_AVAILABLE(executor->ee_type);
> ```
> 
> Note, the condition passed during initialization when creating the cpu_executor. Consequently, I get some warnings at the end along the lines of
> 
> ```
>  mpool.c:54   UCX  WARN  object 0x555d77922f80 was not returned to mpool ec cpu executors
> ```
> 
> which I think is a direct consequence of this, since the mpool_put() of the executor which is done during ee_executor_finalize() is not performed.
> 
> Do you have a suggestion on how to deal with this?

@edgargabriel yes, there is no guarantee ec/cpu will be closed after ec/rocm. What i meant, is that you would call ucc_ee_executor_init(type=CPU) + ucc_ee_executor_start inside your ucc_ec_rocm_executor_init , NOT from static ec_rocm component init. Then you would also do executor_stop + finalize (cpu) in your ucc_ee_rocm_executor_finalize. This way you don't have the issue above. So, instead of having just 1 cpu executor you init/finalize it each time you need to init/finalize rocm executor. Those calls to executor init/start stop/finalize in ec/cpu are very cheap - they mostly do nothing, just taking executor object from pool.

File: src/components/ec/rocm/ec_rocm_executor_interruptible.c [hunk 2]
```diff
@@ -9,6 +9,60 @@
...
+
+    if (task_args->task_type == UCC_EE_EXECUTOR_TASK_REDUCE) {
+        dt     = task_args->reduce.dt;
+    } else {
+        ucc_assert(task->task_type == UCC_EE_EXECUTOR_TASK_REDUCE_STRIDED);
+        dt     = task_args->reduce_strided.dt;
+    }
+    if (dt != UCC_DT_BFLOAT16        &&
+        dt != UCC_DT_FLOAT16         &&
+        dt != UCC_DT_FLOAT32_COMPLEX &&
```
[the remark this answers is not in the document]
Comment (edgargabriel) [reply src/components/ec/rocm/ec_rocm_executor_interruptible.c:59]: I wanted to make the concept work on the standard datatypes first, can revisit the topic to support complex datatypes in a bit

**Comment by edgargabriel** [comment]
@vspetrov thank you, I can confirm that making this change resolved the issue. I think the code is now ready for review again.

**Review by vspetrov — APPROVED** [review]

**Comment by edgargabriel** [comment]
I rebased the branch to master, would it be possible to try to get this merged this week?

File: src/components/ec/rocm/ec_rocm.c [hunk 3]
```diff
@@ -75,6 +75,19 @@ static ucc_config_field_t ucc_ec_rocm_config_table[] = {
      ucc_offsetof(ucc_ec_rocm_config_t, reduce_num_blocks),
      UCC_CONFIG_TYPE_ULUNITS},
 
+     {"REDUCE_HOST_LIMIT", "256",
```
[the remark this answers is not in the document]
Comment (edgargabriel) [reply src/components/ec/rocm/ec_rocm.c:78]: done, thank you for the hint!

File: src/components/ec/rocm/ec_rocm_executor_interruptible.c [hunk 4]
```diff
@@ -7,8 +7,78 @@
 
 #include "ec_rocm_executor.h"
 #include "components/mc/ucc_mc.h"
+#include "components/ec/ucc_ec.h"
 #include "utils/ucc_atomic.h"
 
+static bool ucc_ec_rocm_copy_multi_use_host (const ucc_ee_executor_task_args_t* task_args)
+{
+    bool result = true;
```
[the remark this answers is not in the document]
Comment (edgargabriel) [reply src/components/ec/rocm/ec_rocm_executor_interruptible.c:15]: I don't mind changing it, but bool is part of the C language starting from C99. I would be completely shocked to find a compiler that does not support the bool type at this point in time.

**Review by Sergei-Lebedev — APPROVED** [review]

**Comment by vspetrov** [comment]
bot:retest

**Pull request merged**